drivers/sensor/st/stm32_digi_temp/stm32_digi_temp.c

```diff
@@ -0,0 +1,286 @@
+	 * Allowed values are between 0 and 127.
+	 */
+	clk_div = MIN(DIV_ROUND_UP(data->pclk_freq, ONE_MHZ), 127);
+	MODIFY_REG(dts->CFGR1, DTS_CFGR1_HSREF_CLK_DIV_Msk,
+		   clk_div << DTS_CFGR1_HSREF_CLK_DIV_Pos);
+
+	/* Select PCLK as reference clock */
+	MODIFY_REG(dts->CFGR1, DTS_CFGR1_REFCLK_SEL_Msk,
+		   0 << DTS_CFGR1_REFCLK_SEL_Pos);
+
+	/* Select trigger */
+	MODIFY_REG(dts->CFGR1, DTS_CFGR1_TS1_INTRIG_SEL_Msk,
+		   0 << DTS_CFGR1_TS1_INTRIG_SEL_Pos);
+
+	/* Set sampling time */
+	MODIFY_REG(dts->CFGR1, DTS_CFGR1_TS1_SMP_TIME_Msk,
+		   SAMPLING_TIME << DTS_CFGR1_TS1_SMP_TIME_Pos);
+}
+
+static void stm32_digi_temp_enable(const struct device *dev)
+{
+	const struct stm32_digi_temp_config *cfg = dev->config;
+	DTS_TypeDef *dts = cfg->base;
+
+	/* Enable the sensor */
+	SET_BIT(dts->CFGR1, DTS_CFGR1_TS1_EN);
+
+	/* Enable interrupt */
+	SET_BIT(dts->ITENR, DTS_ITENR_TS1_ITEEN);
+}
+
+#ifdef CONFIG_PM_DEVICE
+static void stm32_digi_temp_disable(const struct device *dev)
+{
+	const struct stm32_digi_temp_config *cfg = dev->config;
+	DTS_TypeDef *dts = cfg->base;
+
+	/* Disable interrupt */
+	CLEAR_BIT(dts->ITENR, DTS_ITENR_TS1_ITEEN);
+
+	/* Disable the sensor */
+	CLEAR_BIT(dts->CFGR1, DTS_CFGR1_TS1_EN);
+}
+#endif
+
+static int stm32_digi_temp_init(const struct device *dev)
+{
+	const struct stm32_digi_temp_config *cfg = dev->config;
+	struct stm32_digi_temp_data *data = dev->data;
+	DTS_TypeDef *dts = cfg->base;
+
+	/* enable clock for subsystem */
+	const struct device *const clk = DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE);
+
+	if (!device_is_ready(clk)) {
+		LOG_ERR("Clock control device not ready");
+		return -ENODEV;
+	}
+
+	if (clock_control_on(clk, (clock_control_subsys_t) &cfg->pclken) != 0) {
+		LOG_ERR("Could not enable DTS clock");
+		return -EIO;
+	}
+
+	/* Save the peripheral clock frequency in the data structure to avoid
+	 * querying it for each call to the channel_get method.
+	 */
+	if (clock_control_get_rate(clk, (clock_control_subsys_t) &cfg->pclken,
+				   &data->pclk_freq) < 0) {
+		LOG_ERR("Failed call clock_control_get_rate(pclken)");
+		return -EIO;
+	}
+
+	/* Save the calibration data in the data structure to avoid reading
+	 * them for each call to the channel_get method, as this requires
+	 * enabling the peripheral clock.
+	 */
+	data->ramp_coeff = dts->RAMPVALR & DTS_RAMPVALR_TS1_RAMP_COEFF;
+	data->t0_freq = (dts->T0VALR1 & DTS_T0VALR1_TS1_FMT0) * 100; /* 0.1 kHz -> Hz */
+
+	/* T0 temperature from the datasheet */
+	switch (dts->T0VALR1 >> DTS_T0VALR1_TS1_T0_Pos) {
+	case 0:
+		data->t0 = TS1_T0_VAL0;
+		break;
+	case 1:
+		data->t0 = TS1_T0_VAL1;
+		break;
+	default:
+		LOG_ERR("Unknown T0 temperature value");
+		return -EIO;
+	}
+
+	/* Init mutex and semaphore */
+	k_mutex_init(&data->mutex);
+	k_sem_init(&data->sem_isr, 0, 1);
+
+	/* Configure interrupt */
+	IRQ_CONNECT(
+		DT_INST_IRQN(0),
+		DT_INST_IRQ(0, priority),
+		stm32_digi_temp_isr,
+		DEVICE_DT_INST_GET(0),
+		0);
+	irq_enable(DT_INST_IRQN(0));
```

Comment:
We shouldn't hard code an instance number within a function that takes a `struct device *` as an argument